### ISSUE TITLE
Excessive Database Calls and Bad Performance

### ISSUE THREAD

**Issue by neo302** [nopSolutions/nopCommerce#5013]
nopCommerce version: Dev

Steps to reproduce the problem: 
The current dev version as of Sept 4th has very bad performance problems from what I'm seeing compared to early versions of 4.3.  I'm using SQL Server and noticed an excessive amount of DB calls via a trace.  The current dev version is so slow, it is not even usable. It is not the database as when I rollback to the old 4.3 code, it is fast as expected.  

Just by simply clicking a category or different menu items, I can watch the amount of CPU used by SQL Server spike up to 50% at times.  I've done everything in the performance guide as well and it didn't make a difference.  **Every call is slow**, as in 15+ seconds slow.  This is with running the vanilla source code on the server.  I am running Nopcommerce as a child website and using https.  The store is tiny with a couple products.

I think the slowness got the worst with the updates around August 27th. 

Some of these queries are called numerous times a second. This query is called the most. 
SELECT
	[t1].[Id],
	[t1].[StoreId],
	[t1].[Value],
	[t1].[Name]
FROM
	[Setting] [t1]
ORDER BY
	[t1].[Name],
	[t1].[StoreId]

I've noticed high overall cpu on the server as well with it hitting 100% and staying there for quite some time.  
I have no warnings on the dashboard related to errors.

Here is some output from a remote debug:
Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executing endpoint 'Nop.Web.Controllers.ScheduleTaskController.RunTask (Nop.Web)'
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker: Information: Route matched with {action = "RunTask", controller = "ScheduleTask", area = ""}. Executing controller action with signature Microsoft.AspNetCore.Mvc.IActionResult RunTask(System.String) on controller Nop.Web.Controllers.ScheduleTaskController (Nop.Web).
Microsoft.AspNetCore.Mvc.StatusCodeResult: Information: Executing HttpStatusCodeResult, setting HTTP status code 204
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker: Information: Executed action Nop.Web.Controllers.ScheduleTaskController.RunTask (Nop.Web) in 42.462ms
Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executed endpoint 'Nop.Web.Controllers.ScheduleTaskController.RunTask (Nop.Web)'
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished in **7434.2656ms** 204 
System.Net.Http.HttpClient.default.ClientHandler: Information: Received HTTP response after 7446.5324ms - NoContent
System.Net.Http.HttpClient.default.LogicalHandler: Information: End processing HTTP request after 7455.2289ms - NoContent
The thread 0x2f28 has exited with code 0 (0x0).
The thread 0x2564 has exited with code 0 (0x0).
Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor: Information: Executed ViewResult - view Index executed in 52326.0635ms.
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker: Information: Executed action Nop.Web.Areas.Admin.Controllers.HomeController.Index (Nop.Web) in **59481.749ms**
Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executed endpoint 'Nop.Web.Areas.Admin.Controllers.HomeController.Index (Nop.Web)'
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished in **65432.6648ms** 200 text/html; charset=utf-8

Thanks

**Comment by skoshelev** [comment]
I have not been able to replicate your problem on the current source code, it works fine now. In your case, there may have been some kind of failure with the caching settings, please check if you have set the CacheConfig settings in the appsettings.json file

Closed #5013

**Comment by kpmarlin** [comment]
@neo302 I'm having the same exact issue on Azure.  Did you ever get this resolved?

@skoshelev Can you post an example of CacheConfig settings?  Mine is below.

```
{
  "Hosting": {
    //Set to "true" the settings below if your hosting uses a load balancer. It'll be used to determine whether the current request is HTTPS
    "UseHttpClusterHttps": false,
    //Also set to "true" the next settings if you use a reverse proxy server (for example, if you host your site on Linux with Nginx/Apache and SSL) 
    "UseHttpXForwardedProto": false,

    //Use the setting below if your hosting doesn't use "X-FORWARDED-FOR" header to determine IP address.
    //In some cases server use other HTTP header. You can specify a custom HTTP header here. For example, CF-Connecting-IP, X-FORWARDED-PROTO, etc
    "ForwardedHttpHeader": ""
  },

  "Nop": {
    //Enable if you want to see the full error in production environment. It's ignored (always enabled) in development environment
    "DisplayFullErrorStack": false,

    //Windows Azure BLOB storage.
    //Specify your connection string, container name, end point for BLOB storage here
    "AzureBlobStorageConnectionString": "",
    "AzureBlobStorageContainerName": "",
    "AzureBlobStorageEndPoint": "",
    "AzureBlobStorageAppendContainerName": "true",

    //Windows Azure BLOB storage for Data Protection Keys (the UseRedisToStoreDataProtectionKeys option should be disabled)
    "UseAzureBlobStorageToStoreDataProtectionKeys": "false",
    //This should be a _PRIVATE_ container separate from the blob container used for media storage
    "AzureBlobStorageContainerNameForDataProtectionKeys": "",
    //Optionally encrypt the Data Protection Keys using an Azure Key Vault
    "AzureKeyVaultIdForDataProtectionKeys": "",

    //Redis support (used by web farms, Azure, etc). Find more about it at https://azure.microsoft.com/en-us/documentation/articles/cache-dotnet-how-to-use-azure-redis-cache/
    "RedisEnabled": false,
    //Redis database id; If you need to use a specific redis database, just set its number here. Set empty if should use the different database for each data type (used by default); set -1 if you want to use the default database
    "RedisDatabaseId": "",
    "RedisConnectionString": "127.0.0.1:6379,ssl=False",
    "UseRedisToStoreDataProtectionKeys": false,
    "UseRedisForCaching": false,
    //Should we ignore Redis timeout exception? Enabling this setting increases Redis cache stability but may decrease site performance
    "IgnoreRedisTimeoutException": false,
    "UseRedisToStorePluginsInfo": false,

    //You can get the latest version of user agent strings at http://browscap.org/
    //Leave "CrawlersOnlyDatabasePath" attribute empty if you want to use full version of "browscap.xml" file
    "UserAgentStringsPath": "~/App_Data/browscap.xml",
    "CrawlerOnlyUserAgentStringsPath": "~/App_Data/browscap.crawlersonly.xml",

    //Do not edit this element. For advanced users only
    "DisableSampleDataDuringInstallation": false,
    "PluginsIgnoredDuringInstallation": "",

    //Enable if you want to clear /Plugins/bin directory on application startup
    "ClearPluginShadowDirectoryOnStartup": true,
    //Enable if you want to copy "locked" assemblies from /Plugins/bin directory to temporary subdirectories on application startup
    "CopyLockedPluginAssembilesToSubdirectoriesOnStartup": false,
    //Enable if you want to copy plugins library to the /Plugins/bin directory on application startup
    "UsePluginsShadowCopy": true,
    //Enable if you want to load an assembly into the load-from context, by passing some security checks
    "UseUnsafeLoadAssembly": true,

    //Configures the wait time (in seconds) before terminating the attempt to execute a command and generating an error.
    //Set null to use default value (30 seconds)
    "SQLCommandTimeout": null,

    //Enable to store TempData in the session state
    "UseSessionStateTempDataProvider": false
  }
}
```

**Comment by neo302** [comment]
@kpmarlin I did not get it resolved as I couldn't spend anymore time on it.  I went back to version 4.3.

**Comment by kpmarlin** [comment]
Hello @neo302.  Thanks for the follow up.  I was able to get this resolved.  I was able to narrow down the error to NopCommerce's Anywhere Sliders plugin.  For every store and every slider on the page, it would update a setting:

```
exec sp_executesql N'UPDATE
[t1]
SET
[t1].[StoreId] = @StoreId,
[t1].[Value] = @Value0,
[t1].[Name] = @Name
FROM
[Setting] [t1]
WHERE
[t1].[Id] = @Id
',N'@StoreId int,@Value0 nvarchar(4000),@Name nvarchar(200),@Id int',@StoreId=0,@Value0=N'247',@Name=N'anywhereslidersettings.customcssfileversion',@Id=2633
```

which would then cause nopCommerce to update settings:

```
SELECT
[t1].[Id],
[t1].[StoreId],
[t1].[Value],
[t1].[Name]
FROM
[Setting] [t1]
ORDER BY
[t1].[Name],
[t1].[StoreId]
```

After contacting nop-templates.com, they said the issue had been fixed in version 4.32 of that plugin.  Assuming you're running that plugin, updating that plugin will fix the issue.  Hope this helps when performing your next upgrade!

**Comment by neo302** [comment]
@kpmarlin Thank you!  I will keep this in mind when I upgrade!
